Calculate throttle thresholds better

kgaillot opened the pull request:
This fixes multiple issues with crmd's throttle checks, most importantly: changes in CPU count due to hot-plugging are now detected; blocked processes are correctly parsed from /proc/stat; and I/O thresholds are set reasonably.

jnpkrn commented on lib/common/procfs.c:
```diff
@@ -140,3 +140,35 @@ crm_procfs_pid_of(const char *name)
+            if (strncmp(buffer, "processor", 9) == 0) {
+                ++*cores;
+            }
+        }
```
While at it, I know it's just copy-paste, but this is pretty naive way to look for a _contiguous_ string in a _contiguous_ block because of parcelling the windows.

jnpkrn replied:
And it's not a single occurrence of this best-wishes antipattern.

jnpkrn replied:
Ah, sorry, refreshed my memory about fgets.

jnpkrn replied:
So the only thing for consideration is that buffer is excessively big for the task.

kgaillot replied:
On my system, the longest line in /proc/cpuinfo is 558 bytes, so actually I think I'll make it larger. The code will work whatever the size, but lines bigger than the buffer will require additional reads, which are typically more expensive than using an extra <1KB memory.

jnpkrn replied:
Yes, gave it some more trial.  What we really need here is:
- read 9 chars from the beginning of the line and compare
- skip until the end of line if not present in those 9 chars already
- repeat
`fgets` does not handle the middle part for us, so it would have to be added by hand.

kgaillot replied:
Skipping to the end of the line is handled by reading up to a newline, which is what fgets does.

jnpkrn replied:
True, yeah, increasing the buffer makes sense now (I have the longest
line 701 chars).

jnpkrn replied:
(This time I was confused with the plain old approach vs.
mmap'ing file without copying something somewhere needlessly
~ array access approach).

jnpkrn commented:
Noticed that `throttle_fini` is either superfluous or should be called on exit for extra sanity (in valgrind etc.).

jnpkrn commented on crmd/throttle.c:
```diff
@@ -102,14 +83,16 @@ int throttle_num_cores(void)
  *       This will return NULL if the daemon is being run via valgrind.
```
... or if `cib` is run under different user/group than `crmd` (unless the latter runs atypically as root)

?

jnpkrn replied:
... under some circumstances

(https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux-stable.git/diff/Documentation/filesystems/proc.txt?id=0499680a42141d86417a8fbaa8c8db806bea1201&id2=97412950b10e64f347aec4a9b759395c2465adf6)

kgaillot replied:
cib and crmd both run as hacluster. We definitely depend on /proc being mounted with permissive options though. There's a similar problem for OSes without /proc as well. It's rare enough that it's up to users with that use case to submit a feature request and/or patch.

jnpkrn replied:
Wanted to highlight that's nothing granted for non-root processes
and we should deal with it.

It's actually quite sane not prevent users inspection of foreign
processes, it's not their business.

jnpkrn replied:
Also note that with permissive options, there's another undesired
side-effect: unreleated user on the machine can do:

```
> cib cat <<EOF
#!/bin/sh
while true; do factor ${RANDOM} >/dev/null; done
EOF
chmod +x cib
./cib
```

and possibly influence the run of the whole cluster, which should not
happen, at least not so easily.

That's all due to the clunky code to get details about running CIB
daemon.  If anything else, pacemaker (mcp) knows very specific PIDs
to look at, and also exports its own API that could be perhaps used
to query the PID (or info about the process required).

kgaillot commented:
Updated:
- fixed commit message
- increased cpuinfo buffer to 1024
- added new commit fixing i/o load calculation

I'll handle throttle_fini() later since it's not essential and would cause a conflict with another pending PR.

jnpkrn commented:
`fini` thing is fine when postponed, the remark was added because without
this PR, I don't think I would notice.

kgaillot commented:
Complete revamp, it was determined that I/O load is not necessary to check separately.
